### PR TITLE
Cloudant search index query

### PR DESCRIPTION
**Pull request by emlaver**
## What

Support Cloudant search query parameter processing and execution.

## How

- Handled search query parameter processing and execution in `CloudantDatabase.get_search_result`
- Added dictionary of search query parameters in common_util

## Testing

- Unit tests for search index query in `CloudantDatabaseTests` class
- Created a search index during test class setup for querying tests with `create_search_index`

## Reviewers

reviewer: @alfinkel 
reviewer @ricellis

## Issues

#158

### CODE REVIEW

File: src/cloudant/database.py [hunk 1]
```diff
@@ -1098,3 +1101,123 @@ def get_query_result(self, selector, fields=None, raw_result=False,
             return QueryResult(query, **kwargs)
         else:
             return query.result
+
+    def get_search_result(self, ddoc_id, index_name, query, **kwargs):
+        """
+        Retrieves the raw JSON content from the remote database based on the
+        search index on the server, using the kwargs provided as query
+        parameters.
+
```
Comment (alfinkel) [inline src/cloudant/database.py]: You mention using kwargs as query parameters.  You might also want to mention using `query` as well.  Just a thought.

File: tests/unit/database_tests.py [hunk 2]
```diff
@@ -1065,5 +1066,108 @@ def test_get_query_indexes(self):
...
+    def test_get_search_result_with_invalid_argument(self):
+        """
+        Test get_search_result by passing in invalid arguments
+        """
+        try:
+            self.db.get_search_result('searchddoc001', 'searchindex001',
+                                      'julia*', foo={'bar': 'baz'})
+            self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError as err:
+            self.assertEqual(str(err), 'Invalid argument: foo')
```
Comment (alfinkel) [inline tests/unit/database_tests.py]: Rather than using the try/except block to test invalid conditions you should use `assertRaises`.  See this #80 for detail.  This applies to all new tests that you added as part of this PR.

File: tests/unit/database_tests.py [hunk 3]
```diff
@@ -1065,5 +1066,108 @@ def test_get_query_indexes(self):
...
+            limit=5,
+            include_docs=True
+        )
+        self.assertEqual(len(resp['rows']), 5)
+        self.assertEqual(resp['total_rows'], 100)
+        for row in resp['rows']:
+            self.assertIsNotNone(row['fields'])
+            self.assertTrue(row['id'].startswith('julia0'))
+            self.assertIsNotNone(row['order'])
+            self.assertIsNotNone(row['doc'])
```
Comment (alfinkel) [inline tests/unit/database_tests.py]: I don't really like stepping through the elements of the `resp` and performing asserts inside of the loop.  Would it be possible to compose an object that looks like what you expect the `resp` to look like based on the query call and then to use that object to do a direct comparison to the `resp`?  This is was what I've tried to do as much as possible in this library and what most of the other client libraries also attempt to do.  Same comment for the group_option test below.

File: pylintrc [hunk 4]
```diff
@@ -235,7 +235,7 @@ notes=FIXME,XXX,TODO
 [SIMILARITIES]
 
 # Minimum lines number of a similarity.
-min-similarity-lines=5
+min-similarity-lines=10
```
Comment (alfinkel) [inline pylintrc]: Is this change necessary any longer not that we've remove a lot of the SearchIndex stuff?

File: tests/unit/database_tests.py [hunk 5]
```diff
@@ -640,6 +640,7 @@ def setUp(self):
         """
         super(CloudantDatabaseTests, self).setUp()
         self.db_set_up()
+        self.create_search_index()
```
Comment (alfinkel) [inline tests/unit/database_tests.py]: Is it worth having this in the `setUp` for 6 tests?  Maybe move it to the tests themselves?

File: src/cloudant/database.py [hunk 6]
```diff
@@ -1098,3 +1101,123 @@ def get_query_result(self, selector, fields=None, raw_result=False,
...
+        :param int highlight_size: A number of characters in each fragment for
+            highlights.  Defaults to 100 characters.
+        :param list include_fields: A list of field names to include in search
+            results. Any fields included must have been indexed with the
+            ``store:true`` option.
+
+        :returns: Search query result data in JSON format
+        """
+        data = dict(query=query)
+        data.update(kwargs)
```
Comment (alfinkel) [inline src/cloudant/database.py]: Unfortunately, I believe that this creates a vulnerability where someone can provide a `query` as part of the `kwargs` and basically overwrite the `query` provided as a separate argument.  This could cause a lot of confusion for a developer that mistakenly falls into this trap.

I think it would be cleaner/safer to change the method signature to `def get_search_result(self, doc_id, index_name, **query_params)` and to then perform a check for `query` since it is mandatory.  As in:

```python
def get_search_result(self, doc_id, index_name, **query_params):
    ....
    if not query_params.get('query'):
        # raise CloudantArgumentError exception
    data = dict(query_params)
    ....
```

File: src/cloudant/database.py [hunk 7]
```diff
@@ -1098,3 +1101,123 @@ def get_query_result(self, selector, fields=None, raw_result=False,
...
+                    'Argument {0} is not an instance of expected type: {1}'
+                ).format(key, SEARCH_INDEX_ARGS[key])
+                raise CloudantArgumentError(msg)
+        # Execute query search
+        headers = {'Content-Type': 'application/json'}
+        ddoc = DesignDocument(self, ddoc_id)
+        resp = self.r_session.post(
+            '/'.join([ddoc.document_url, '_search', index_name]),
+            headers=headers,
+            data=json.dumps(data, cls=self.client.encoder)
```
Comment (alfinkel) [inline src/cloudant/database.py]: As discussed over Slack, you no longer need `data` since you are not manipulating the query parameters any longer.  You can get rid of the `data` variable and use `query_params` directly here.

File: src/cloudant/database.py (hunk 1 above)
Comment (emlaver) [inline src/cloudant/database.py]: Changed method signature and removed `query` argument in 8e22bed.

File: tests/unit/database_tests.py (hunk 2 above)
Comment (emlaver) [inline tests/unit/database_tests.py]: Fixed in 8e22bed.

File: tests/unit/database_tests.py (hunk 3 above)
Comment (emlaver) [inline tests/unit/database_tests.py]: I've composed expected `resp` objects in 8e22bed.

File: pylintrc (hunk 4 above)
Comment (emlaver) [inline pylintrc]: I've changed `min-similarity-lines` back to 5 in 8e22bed.

File: tests/unit/database_tests.py (hunk 5 above)
Comment (emlaver) [inline tests/unit/database_tests.py]: Only two unit tests require `create_search_index` so I've moved it in 8e22bed.

File: src/cloudant/database.py (hunk 6 above)
Comment (emlaver) [inline src/cloudant/database.py]: Changed method signature with updates to docstring in 8e22bed.

File: src/cloudant/database.py (hunk 7 above)
Comment (emlaver) [inline src/cloudant/database.py]: Fixed in 8e22bed.

**Comment by alfinkel** [comment]
+1

File: src/cloudant/database.py [hunk 8]
```diff
@@ -1098,3 +1101,128 @@ def get_query_result(self, selector, fields=None, raw_result=False,
...
+            optional and defaults to number.
+        :param str stale: Optional string to allow the results from a stale
+            index to be used. This makes the request return immediately, even
+            if the index has not been completely built yet.
+        :param list highlight_fields: Optional list of fields which should be
+            highlighted.
+        :param str highlight_pre_tag: Optional string inserted before the
+            highlighted word in the highlights output.  Defaults to ``<em>``.
+        :param str highlight_post_tag: Optional string inserted after the
+            highlighted word in the highlights output.  Defaults to ``<em>``.
```
Comment (ricellis) [inline src/cloudant/database.py]: Shouldn't this be `</em>`?

File: src/cloudant/database.py [hunk 9]
```diff
@@ -1098,3 +1101,128 @@ def get_query_result(self, selector, fields=None, raw_result=False,
...
+                'containing Lucene syntax and retry.')
+
+        # Validate query arguments and values
+        for key, val in iteritems_(query_params):
+            if key not in list(SEARCH_INDEX_ARGS.keys()):
+                msg = 'Invalid argument: {0}'.format(key)
+                raise CloudantArgumentError(msg)
+            if not isinstance(val, SEARCH_INDEX_ARGS[key]):
+                msg = (
+                    'Argument {0} is not an instance of expected type: {1}'
```
Comment (ricellis) [inline src/cloudant/database.py]: This message could be a little confusing for the cases where we accept multiple types. For instance `query` and `sort`. If query was the wrong type for example I think the message would read:
>Argument query is not an instance of expected type: `(<type 'basestring'>, <type 'int'>)`

I can see why is preferable to do this check using the tuple as it makes the code much simpler. I'm not sure saying "one of the expected types:" is really any better, but it might be a bit less ambiguous:
>Argument query is not an instance of one of the expected types: `(<type 'basestring'>, <type 'int'>)`

although this looks a bit weird in the singular case:
>Argument stale is not an instance of one of the expected types: `(<type 'basestring'>)`

File: tests/unit/database_tests.py [hunk 10]
```diff
@@ -1065,5 +1065,156 @@ def test_get_query_indexes(self):
...
+    def test_get_search_result_with_invalid_value_types(self):
+        """
+        Test get_search_result by passing in invalid value types for
+        query parameters
+        """
+        test_data = [
+            {'bookmark': 1},                    # Should be a basestring
+            {'counts': 'blah'},                 # Should be a list
+            {'drilldown': 'blah'},              # Should be a list
+            {'group_field': ['blah']},          # Should be an basestring
```
Comment (ricellis) [inline tests/unit/database_tests.py]: Looks like you missed a couple on copy/paste
a basestring

File: tests/unit/database_tests.py [hunk 11]
```diff
@@ -1065,5 +1065,156 @@ def test_get_query_indexes(self):
...
+            {'group_limit': 'int'},             # Should be an int
+            {'group_sort': 'blah'},             # Should be a list
+            {'include_docs': 'blah'},           # Should be an boolean
```
Comment (ricellis) [inline tests/unit/database_tests.py]: a boolean

File: src/cloudant/database.py (hunk 9 above)
Comment (emlaver) [inline src/cloudant/database.py]: Looks like it could be more common for a user to see the singular case knowing that 6 of the 18 query parameters accept multiple types.  If we change this instance of the message to `an instance of one of the expected types`, we would also want to consider changing the other instances in the project.

File: src/cloudant/database.py (hunk 8 above)
Comment (emlaver) [inline src/cloudant/database.py]: This default value was copied directly from the documentation:
https://docs.cloudant.com/search.html#queries

File: tests/unit/database_tests.py (hunk 10 above)
Comment (emlaver) [inline tests/unit/database_tests.py]: Fixed in 4e706c2.

File: tests/unit/database_tests.py (hunk 11 above)
Comment (emlaver) [inline tests/unit/database_tests.py]: Fixed in 4e706c2.

File: src/cloudant/database.py (hunk 9 above)
Comment (ricellis) [inline src/cloudant/database.py]: Yeah I see that the singular is more common, but to my mind even the slightly weird singular case:
>Argument stale is not an instance of one of the expected types: `(<type 'basestring'>)`

makes more sense than the reverse:
>Argument query is not an instance of expected type: `(<type 'basestring'>, <type 'int'>)`

I thought most of the other places in the codebase were singular, except when ints are expected so it is `(int, NONETYPE)`.
I guess an option would be to replace all the places these messages are generated with calls to a single utility method to format the message based on the number of types checked. If that is a better option I think that could happen under another issue/PR.

File: src/cloudant/database.py (hunk 8 above)
Comment (ricellis) [inline src/cloudant/database.py]: Surely that doc must be incorrect `<em>match<em>` would not be valid.

File: src/cloudant/database.py (hunk 9 above)
Comment (emlaver) [inline src/cloudant/database.py]: I think it's better to handle this under another issue.

Comment (alfinkel) [inline src/cloudant/database.py]: btw - `basestring` does not exist in Python 3.

Comment (emlaver) [inline src/cloudant/database.py]: Thanks @alfinkel. From the comment above, it's probably best to change the comments with `basestring` to:
```
# Should be a STRTYPE
```

File: src/cloudant/database.py (hunk 8 above)
Comment (emlaver) [inline src/cloudant/database.py]: You're correct, it's invalid.  Result from a quick curl test:
```
{..., "rows":[...,"highlights":{"name":["<em>aardvark</em>"]}}]}
```
Fixed in 90213c1.

File: tests/unit/database_tests.py [hunk 12]
```diff
@@ -1065,5 +1065,156 @@ def test_get_query_indexes(self):
...
+        err = cm.exception
+        self.assertEqual(str(err), 'Invalid argument: foo')
+
+    def test_get_search_result_with_invalid_value_types(self):
+        """
+        Test get_search_result by passing in invalid value types for
+        query parameters
+        """
+        test_data = [
+            {'bookmark': 1},                    # Should be a basestring
```
Comment (ricellis) [inline tests/unit/database_tests.py]: So is the plan to change these comments to say `a STRTYPE` now?

**Comment by ricellis** [comment]
OK so I think we're agreed:

1. Open a new issue for replacing the type checking messages.
2. Fix the comments referring to `basestring` to `STRTYPE` to suit both python versions

+1

**Comment by emlaver** [comment]
@ricellis Issue opened: https://github.com/cloudant/python-cloudant/issues/176
